File: Core/src/main/java/com/hqq/core/toolbar/BaseDefToolBarImpl.java
package com.hqq.core.toolbar;

import android.app.Activity;

import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.DrawableRes;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.hqq.core.R;
import com.hqq.core.utils.ResourcesUtils;
import com.hqq.core.widget.FilterImageView;

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.toolbar
 * @FileName :   BaseDefToolBarImpl
 * @Date : 2018/11/15 0015  下午 7:27
 * @Descrive :
 * @Email :  qiqiang213@gmail.com
 */
public class BaseDefToolBarImpl extends BaseToolBar {
    private ImageView mImageViewBg;
    private int mDefToolBarColor = R.color.toolbar_bg_color;
    private int mDefTitleColor = R.color.color_333;

    @Override
    public View iniToolBar(final Activity activity, ViewGroup viewGroup) {
        Toolbar toolbar = (Toolbar)
                LayoutInflater.from(activity.getBaseContext()).inflate(R.layout.layout_def_toolbar, viewGroup, false);
        if (activity instanceof AppCompatActivity) {
            ((AppCompatActivity) activity).setSupportActionBar(toolbar);
        }
        toolbar.findViewById(R.id.iv_bar_back).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                activity.onBackPressed();
            }
        });
        mImageViewBg = toolbar.findViewById(R.id.iv_toolBar_Bg);
        mImageViewBg.setBackgroundResource(mDefToolBarColor);
        ((TextView) toolbar.findViewById(R.id.tv_bar_title)).setText(activity.getTitle());
        ((TextView) toolbar.findViewById(R.id.tv_bar_title)).setTextColor(ContextCompat.getColor(toolbar.getContext(), mDefTitleColor));
        return toolbar;
    }


    /**
     * 滑动效果
     *
     * @param ahpla 0 -1
     */
    @Override
    public void initScroll(float ahpla) {
        if (ahpla > 1) {
            ahpla = 1;
        }
        super.initScroll(ahpla);
        if (ahpla < 0.5) {
            getLeftView().setImageResource(R.mipmap.ic_black_whit);
            getLeftView().setAlpha(1 - ahpla * 2);
        } else {
            getLeftView().setImageResource(R.mipmap.ic_back_gray);
            getLeftView().setAlpha((float) ((ahpla - 0.5) * 2));
        }

        if (getToolbarTitle() != null) {
            getToolbarTitle().setAlpha(ahpla);
        }
        if (mImageViewBg != null) {
            mImageViewBg.setAlpha(ahpla);
        }

    }

    @Override
    public void setToolBarColor(int color) {
        if (null != getToolBarBg()) {
            getToolBarBg().setImageResource(color);
        } else {
            mDefToolBarColor = color;
        }
        if (null != getStatusBar()) {
            getStatusBar().setBackgroundResource(color);
        } else {
            mDefStatusColor = color;
        }
    }


    /**
     * 设置背景的透明度
     *
     * @param ahpla
     */
    public void initScrollNoImage(float ahpla) {
        if (ahpla > 1) {
            ahpla = 1;
        }
        super.initScroll(ahpla);
        if (ahpla < 0.5) {
            getLeftView().setAlpha(1 - ahpla * 2);
        } else {
            getLeftView().setAlpha((float) ((ahpla - 0.5) * 2));
        }
        if (getToolbarTitle() != null) {
            getToolbarTitle().setAlpha(ahpla);
        }
        if (mImageViewBg != null) {
            mImageViewBg.setAlpha(ahpla);
        }

    }

    /**
     * 获取左边 图标View
     *
     * @return ImageView
     */
    public ImageView getLeftView() {
        return mRootView.findViewById(R.id.iv_bar_back);
    }


    /**
     * 背景View
     *
     * @returnImageView
     */
    public ImageView getToolBarBg() {
        return mImageViewBg;
    }

    /**
     * 设置 bar 的颜色
     *
     * @param color
     */

    /**
     * 设置标题
     *
     * @param title
     */
    public BaseDefToolBarImpl setToolbarTitle(CharSequence title) {
        if (mRootView == null) {
            return null;
        }
        getToolbarTitle().setText(title);
        return this;
    }

    /**
     * 获取标题View
     *
     * @return
     */
    public TextView getToolbarTitle() {
        return mRootView.findViewById(R.id.tv_bar_title);
    }


    /**
     * 设置背景颜色 需要在view创建完成
     *
     * @param resid
     */
    public BaseDefToolBarImpl setToolBarBg(@DrawableRes int resid) {
        mDefToolBarColor = resid;
        if (mImageViewBg != null) {
            mImageViewBg.setImageResource(resid);
        }
        return this;
    }

    /**
     * 配置默认颜色
     *
     * @param defToolBarColor
     * @return
     */
    public BaseDefToolBarImpl setDefToolBarColor(int defToolBarColor) {
        mDefToolBarColor = defToolBarColor;
        return this;
    }

    /**
     * 设置标题颜色
     *
     * @param defTitleColor
     * @return
     */
    public BaseDefToolBarImpl setDefTitleColor(int defTitleColor) {
        getToolbarTitle().setTextColor(ContextCompat.getColor(getToolbarTitle().getContext(), defTitleColor));
        return this;
    }


    /**
     * 设置 左边图片
     *
     * @param id
     * @return
     */
    public BaseDefToolBarImpl setLetImageView(int id) {
        getLeftView().setImageResource(id);
        return this;
    }

    /**
     * 添加View
     *
     * @param view
     */
    public void addRightView(View view) {
        ((LinearLayout) mRootView.findViewById(R.id.ll_right)).addView(view);
    }

    //region 添加右边文字

    /**
     * 添加文字
     *
     * @param title
     * @param clickListener
     */
    public void addRightTextView(String title, View.OnClickListener clickListener) {
        addRightTextView(title, R.color.color_333, clickListener);
    }

    /**
     * @param title
     * @param color
     * @param clickListener
     */
    public void addRightTextView(String title, @ColorRes int color, View.OnClickListener clickListener) {
        addRightTextView(title, R.color.color_333, R.dimen.x28, clickListener);

    }

    /**
     * @param title
     * @param color
     * @param size
     * @param clickListener
     */
    public void addRightTextView(String title, @ColorRes int color, @DimenRes int size, View.OnClickListener clickListener) {
        addRightView(newTextView(title, color, size, clickListener));
    }

    /**
     * 默认文本
     *
     * @param title
     * @param color
     * @param size
     * @param clickListener
     * @return
     */
    private View newTextView(String title, @ColorRes int color, @DimenRes int size, View.OnClickListener clickListener) {
        TextView textView = new TextView(mRootView.getContext());
        textView.setText(title);
        textView.setPadding(0, 0, (int) ResourcesUtils.getDimen(R.dimen.x20), 0);
        textView.setGravity(Gravity.CENTER);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT);
        textView.setLayoutParams(params);
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, ResourcesUtils.getDimen(size));
        textView.setOnClickListener(clickListener);
        textView.setTextColor(ResourcesUtils.getColor(color));
        return textView;
    }


    //endregion


    /**
     * 添加图片
     *
     * @param icImage
     * @param listener
     */
    public void addRightImageView(@DrawableRes int icImage, View.OnClickListener listener) {
        addRightImageView(icImage, 0, 0, (int) ResourcesUtils.getDimen(R.dimen.x20), 0, listener);
    }

    public void addRightImageView(@DrawableRes int icImage, int left, int top, int right, int bottom, View.OnClickListener listener) {
        addRightView(newImageView(icImage, left, top, right, bottom, listener));
    }

    /**
     * 默认的图片
     *
     * @param icImage
     * @param left
     * @param top
     * @param right
     * @param bottom
     * @param listener
     * @return
     */
    private View newImageView(@DrawableRes int icImage, int left, int top, int right, int bottom, View.OnClickListener listener) {
        FilterImageView imageView = new FilterImageView(mRootView.getContext());
        imageView.setAdjustViewBounds(true);
        imageView.setImageResource(icImage);
        imageView.setScaleType(ImageView.ScaleType.CENTER);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT);
        imageView.setLayoutParams(params);
        imageView.setPadding(left, top, right, bottom);
        imageView.setOnClickListener(listener);
        return imageView;
    }


}

File: Core/src/main/java/com/hqq/core/toolbar/BaseDefToolBarImpl.kt
package com.hqq.core.toolbar

import android.app.Activity
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.DrawableRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import com.hqq.core.R
import com.hqq.core.utils.ResourcesUtils
import com.hqq.core.widget.FilterImageView

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.toolbar
 * @FileName :   BaseDefToolBarImpl
 * @Date : 2018/11/15 0015  下午 7:27
 * @Descrive :
 * @Email :  qiqiang213@gmail.com
 */
class BaseDefToolBarImpl : BaseToolBar() {
    /**
     * 背景View
     *
     * @returnImageView
     */
    var toolBarBg: ImageView? = null
        private set
    private var mDefToolBarColor = R.color.toolbar_bg_color
    private val mDefTitleColor = R.color.color_333
    override fun iniToolBar(activity: Activity, viewGroup: ViewGroup?): View {
        val toolbar = LayoutInflater.from(activity.baseContext).inflate(R.layout.layout_def_toolbar, viewGroup, false) as Toolbar
        if (activity is AppCompatActivity) {
            activity.setSupportActionBar(toolbar)
        }
        toolbar.findViewById<View>(R.id.iv_bar_back).setOnClickListener { activity.onBackPressed() }
        toolBarBg = toolbar.findViewById(R.id.iv_toolBar_Bg)
        toolBarBg.setBackgroundResource(mDefToolBarColor)
        (toolbar.findViewById<View>(R.id.tv_bar_title) as TextView).text = activity.title
        (toolbar.findViewById<View>(R.id.tv_bar_title) as TextView).setTextColor(ContextCompat.getColor(toolbar.context, mDefTitleColor))
        return toolbar
    }

    /**
     * 滑动效果
     *
     * @param ahpla 0 -1
     */
    override fun initScroll(ahpla: Float) {
        var ahpla = ahpla
        if (ahpla > 1) {
            ahpla = 1f
        }
        super.initScroll(ahpla)
        if (ahpla < 0.5) {
            leftView.setImageResource(R.mipmap.ic_black_whit)
            leftView.alpha = 1 - ahpla * 2
        } else {
            leftView.setImageResource(R.mipmap.ic_back_gray)
            leftView.alpha = ((ahpla - 0.5) * 2).toFloat()
        }
        if (toolbarTitle != null) {
            toolbarTitle!!.alpha = ahpla
        }
        if (toolBarBg != null) {
            toolBarBg!!.alpha = ahpla
        }
    }

    override fun setToolBarColor(color: Int) {
        if (null != toolBarBg) {
            toolBarBg!!.setImageResource(color)
        } else {
            mDefToolBarColor = color
        }
        if (null != statusBar) {
            statusBar.setBackgroundResource(color)
        } else {
            mDefStatusColor = color
        }
    }

    /**
     * 设置背景的透明度
     *
     * @param ahpla
     */
    fun initScrollNoImage(ahpla: Float) {
        var ahpla = ahpla
        if (ahpla > 1) {
            ahpla = 1f
        }
        super.initScroll(ahpla)
        if (ahpla < 0.5) {
            leftView.alpha = 1 - ahpla * 2
        } else {
            leftView.alpha = ((ahpla - 0.5) * 2).toFloat()
        }
        if (toolbarTitle != null) {
            toolbarTitle!!.alpha = ahpla
        }
        if (toolBarBg != null) {
            toolBarBg!!.alpha = ahpla
        }
    }

    /**
     * 获取左边 图标View
     *
     * @return ImageView
     */
    val leftView: ImageView
        get() = mRootView.findViewById(R.id.iv_bar_back)

    /**
     * 设置 bar 的颜色
     *
     * @param color
     */
    /**
     * 设置标题
     *
     * @param title
     */
    fun setToolbarTitle(title: CharSequence?): BaseDefToolBarImpl? {
        if (mRootView == null) {
            return null
        }
        toolbarTitle!!.text = title
        return this
    }

    /**
     * 获取标题View
     *
     * @return
     */
    val toolbarTitle: TextView?
        get() = mRootView.findViewById(R.id.tv_bar_title)

    /**
     * 设置背景颜色 需要在view创建完成
     *
     * @param resid
     */
    fun setToolBarBg(@DrawableRes resid: Int): BaseDefToolBarImpl {
        mDefToolBarColor = resid
        if (toolBarBg != null) {
            toolBarBg!!.setImageResource(resid)
        }
        return this
    }

    /**
     * 配置默认颜色
     *
     * @param defToolBarColor
     * @return
     */
    fun setDefToolBarColor(defToolBarColor: Int): BaseDefToolBarImpl {
        mDefToolBarColor = defToolBarColor
        return this
    }

    /**
     * 设置标题颜色
     *
     * @param defTitleColor
     * @return
     */
    fun setDefTitleColor(defTitleColor: Int): BaseDefToolBarImpl {
        toolbarTitle!!.setTextColor(ContextCompat.getColor(toolbarTitle!!.context, defTitleColor))
        return this
    }

    /**
     * 设置 左边图片
     *
     * @param id
     * @return
     */
    fun setLetImageView(id: Int): BaseDefToolBarImpl {
        leftView.setImageResource(id)
        return this
    }

    /**
     * 添加View
     *
     * @param view
     */
    fun addRightView(view: View?) {
        (mRootView.findViewById<View>(R.id.ll_right) as LinearLayout).addView(view)
    }
    //region 添加右边文字
    /**
     * 添加文字
     *
     * @param title
     * @param clickListener
     */
    fun addRightTextView(title: String?, clickListener: View.OnClickListener?) {
        addRightTextView(title, R.color.color_333, clickListener)
    }

    /**
     * @param title
     * @param color
     * @param clickListener
     */
    fun addRightTextView(title: String?, @ColorRes color: Int, clickListener: View.OnClickListener?) {
        addRightTextView(title, R.color.color_333, R.dimen.x28, clickListener)
    }

    /**
     * @param title
     * @param color
     * @param size
     * @param clickListener
     */
    fun addRightTextView(title: String?, @ColorRes color: Int, @DimenRes size: Int, clickListener: View.OnClickListener?) {
        addRightView(newTextView(title, color, size, clickListener))
    }

    /**
     * 默认文本
     *
     * @param title
     * @param color
     * @param size
     * @param clickListener
     * @return
     */
    private fun newTextView(title: String?, @ColorRes color: Int, @DimenRes size: Int, clickListener: View.OnClickListener?): View {
        val textView = TextView(mRootView.context)
        textView.text = title
        textView.setPadding(0, 0, ResourcesUtils.getDimen(R.dimen.x20).toInt(), 0)
        textView.gravity = Gravity.CENTER
        val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
        textView.layoutParams = params
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, ResourcesUtils.getDimen(size))
        textView.setOnClickListener(clickListener)
        textView.setTextColor(ResourcesUtils.getColor(color))
        return textView
    }
    //endregion
    /**
     * 添加图片
     *
     * @param icImage
     * @param listener
     */
    fun addRightImageView(@DrawableRes icImage: Int, listener: View.OnClickListener?) {
        addRightImageView(icImage, 0, 0, ResourcesUtils.getDimen(R.dimen.x20).toInt(), 0, listener)
    }

    fun addRightImageView(@DrawableRes icImage: Int, left: Int, top: Int, right: Int, bottom: Int, listener: View.OnClickListener?) {
        addRightView(newImageView(icImage, left, top, right, bottom, listener))
    }

    /**
     * 默认的图片
     *
     * @param icImage
     * @param left
     * @param top
     * @param right
     * @param bottom
     * @param listener
     * @return
     */
    private fun newImageView(@DrawableRes icImage: Int, left: Int, top: Int, right: Int, bottom: Int, listener: View.OnClickListener?): View {
        val imageView = FilterImageView(mRootView.context)
        imageView.adjustViewBounds = true
        imageView.setImageResource(icImage)
        imageView.scaleType = ImageView.ScaleType.CENTER
        val params = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.MATCH_PARENT)
        imageView.layoutParams = params
        imageView.setPadding(left, top, right, bottom)
        imageView.setOnClickListener(listener)
        return imageView
    }
}